content/en/blog/_posts/2024-09-20-kubeadm-v1beta4.md

```diff
@@ -0,0 +1,99 @@
+---
+layout: blog
+title: 'Kubernetes v1.31: kubeadm v1beta4'
+date: 2024-08-23
+slug: kubernetes-1-31-kubeadm-v1beta4
+author: >
+   Paco Xu (DaoCloud)
+---
+
+The [`kubeadm`](/docs/reference/setup-tools/kubeadm/) configuration
+([v1beta4](/docs/reference/config-api/kubeadm-config.v1beta4/))
+is released with Kubernetes v1.31. v1beta3 is still supported,
+but is marked as deprecated. It will be removed after a minimum
+of 3 Kubernetes releases.
+
+This guide will walk you through key changes and update about the kubeadm v1beta4 configuration,
+and how to migrate from v1beta3 to v1beta4.
+
+Details about v1beta4 can be found in the
+[API doc]((/docs/reference/config-api/kubeadm-config.v1beta4/)).
+The documentation is maintained in the
+[code](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go)
+and generated for the kubernetes.io website.
+
+### A list of changes since v1beta3
+
+This version improves on the [v1beta3](/docs/reference/config-api/kubeadm-config.v1beta3/)
+format by fixing some minor issues and adding a few new fields.
+
+To put it simply,
+
+- Two new API types: ResetConfiguration and UpgradeConfiguration
+- For InitConfiguration and JoinConfiguration, `dryRun` mode and `nodeRegistration.imagePullSerial` are supported
+- For ClusterConfiguration, we add field including `certificateValidityPeriod`,
+`caCertificateValidityPeriod`, `encryptionAlgorithm`, `dns.disabled` and `proxy.disabled`.
+- Support `extraEnvs` for all control plan components
+- `extraArgs` changed from a map to structured extra arguments for duplicates
+- Add a `timeouts` structure for init, join, upgrade and reset.
+
+For details, you can see the official document below:
+
+- Support custom environment variables in control plane components under `ClusterConfiguration`.
+Use `apiServer.extraEnvs`, `controllerManager.extraEnvs`, `scheduler.extraEnvs`, `etcd.local.extraEnvs`.
+- The ResetConfiguration API type is now supported in v1beta4. Users are able to reset a node by passing
+a `--config` file to "kubeadm reset".
+- `dryRun` mode in is now configurable in InitConfiguration and JoinConfiguration.
+- Replace the existing string/string extra argument maps with structured extra arguments that support duplicates.
+The change applies to `ClusterConfiguration` - `apiServer.extraArgs`, `controllerManager.extraArgs`,
+`scheduler.extraArgs`, `etcd.local.extraArgs`. Also to `nodeRegistrationOptions.kubeletExtraArgs`.
```

Comment:
Indent the followup lines in these items would help improve readability and maintainability. That is:

```suggestion
  The change applies to `ClusterConfiguration` - `apiServer.extraArgs`, `controllerManager.extraArgs`,
  `scheduler.extraArgs`, `etcd.local.extraArgs`. Also to `nodeRegistrationOptions.kubeletExtraArgs`.
```